File: java/src/main/java/dev/intermediatebox/array/TwoSum.java
package dev.intermediatebox.array;

import java.util.Arrays;
import java.util.HashMap;

public class TwoSum {
  public static void main(String[] args) {
    Solution s = new Solution();

    System.out.println(Arrays.toString(s.twoSum(new int[]{2, 7, 11, 15},9)));
    System.out.println(Arrays.toString(s.twoSum(new int[]{3, 2, 4}, 6)));
    System.out.println(Arrays.toString(s.twoSum(new int[]{3, 3}, 6)));
  }

  private static class Solution {
    public int[] twoSum(int[] nums, int target) {
      HashMap<Integer, Integer> seen = new HashMap<>();

      for (int i = 0; i < nums.length; i++) {
        int complement = target - nums[i];

        if (seen.containsKey(complement)) {
          return new int[] { i, seen.get(complement)};
        }

        seen.put(nums[i], i);
      }

      throw new IllegalArgumentException("no solution found.");
    }
  }
}

File: typescript/src/array/TwoSum.ts
(() => {
  function twoSum(nums: number[], target: number): number[] {
    const seen = new Map<number, number>();

    for (let i = 0; i < nums.length; i++) {
      const complement = target - nums[i];

      if (seen.has(complement)) {
        return [i, seen.get(complement)!];
      }

      seen.set(nums[i], i);
    }

    throw new Error('no solution found');
  };

  console.log(twoSum([2, 7, 11, 15], 9));
  console.log(twoSum([3, 2, 4], 6));
  console.log(twoSum([3, 3], 6));
})();